src/meta/processors/zone/DivideZoneProcessor.cpp

```diff
@@ -0,0 +1,153 @@
+    onFinished();
+    return;
+  }
+
+  auto& zoneItems = req.get_zone_items();
+  auto zoneHosts = MetaKeyUtils::parseZoneHosts(std::move(nebula::value(zoneValueRet)));
+  if (zoneItems.size() > zoneHosts.size()) {
+    LOG(ERROR) << "Zone Item should not greater than hosts size";
+    handleErrorCode(nebula::cpp2::ErrorCode::E_INVALID_PARM);
+    onFinished();
+    return;
+  }
+
+  std::vector<std::string> zoneNames;
+  std::unordered_set<HostAddr> totalHosts;
+  // std::vector<kvstore::KV> data;
+  auto batchHolder = std::make_unique<kvstore::BatchHolder>();
+  nebula::cpp2::ErrorCode code = nebula::cpp2::ErrorCode::SUCCEEDED;
+  for (auto iter = zoneItems.begin(); iter != zoneItems.end(); iter++) {
+    auto zone = iter->first;
+    auto hosts = iter->second;
+    auto valueRet = doGet(MetaKeyUtils::zoneKey(zone));
+    if (nebula::ok(valueRet)) {
+      LOG(ERROR) << "Zone " << zone << " have existed";
+      code = nebula::cpp2::ErrorCode::E_EXISTED;
+      break;
+    }
+
+    auto it = std::find(zoneNames.begin(), zoneNames.end(), zone);
+    if (it == zoneNames.end()) {
+      LOG(ERROR) << "Zone have duplicated name";
+      zoneNames.emplace_back(zone);
+    } else {
+      code = nebula::cpp2::ErrorCode::E_INVALID_PARM;
+      break;
+    }
+
+    if (hosts.empty()) {
+      LOG(ERROR) << "Hosts should not be empty";
+      code = nebula::cpp2::ErrorCode::E_INVALID_PARM;
+    }
+
+    if (std::unique(hosts.begin(), hosts.end()) != hosts.end()) {
+      LOG(ERROR) << "Zone have duplicated host";
+      code = nebula::cpp2::ErrorCode::E_INVALID_PARM;
+      break;
+    }
+
+    std::copy(hosts.begin(), hosts.end(), std::inserter(totalHosts, totalHosts.end()));
+
+    auto key = MetaKeyUtils::zoneKey(std::move(zone));
+    auto val = MetaKeyUtils::zoneVal(std::move(hosts));
+    batchHolder->put(std::move(key), std::move(val));
+  }
+
+  if (code != nebula::cpp2::ErrorCode::SUCCEEDED) {
+    handleErrorCode(code);
+    onFinished();
+    return;
+  }
+
+  if (totalHosts.size() != zoneHosts.size()) {
+    LOG(ERROR) << "The total host is not all hosts";
+    handleErrorCode(nebula::cpp2::ErrorCode::E_INVALID_PARM);
+    onFinished();
+    return;
+  }
+
+  for (auto& host : totalHosts) {
+    auto iter = std::find(zoneHosts.begin(), zoneHosts.end(), host);
+    if (iter == zoneHosts.end()) {
+      LOG(ERROR) << "Host " << host << " not exist in original zone";
+      code = nebula::cpp2::ErrorCode::E_INVALID_PARM;
+      break;
+    }
+  }
+
+  if (code != nebula::cpp2::ErrorCode::SUCCEEDED) {
+    handleErrorCode(code);
+    onFinished();
+    return;
+  }
+
+  // Remove original zone
+  batchHolder->remove(std::move(zoneKey));
+  code = updateSpacesZone(batchHolder.get(), zoneName, zoneNames);
+  if (code != nebula::cpp2::ErrorCode::SUCCEEDED) {
+    handleErrorCode(code);
+    onFinished();
+    return;
+  }
+
+  auto batch = encodeBatchValue(std::move(batchHolder)->getBatch());
+  doBatchOperation(std::move(batch));
```

Comment:
I have a question:
Why `batch` operation must be used here
Almost all of our meta operations are multi operations

Reply:
because it's used to avoid the operation being interrupted and ensure data consistency

Reply:
Because both put and remove is used

Reply:
👌